IMPORTING LIBRARIES -

1) NumPy is a fundamental package for scientific computing in Python. It provides support for arrays (multi-dimensional), mathematical functions, random number generation, and more.

2) Pandas is a powerful data manipulation and analysis library. It provides data structures like Series (1D) and DataFrame (2D) to handle structured data efficiently.

3) The datasets module in scikit-learn provides utilities to load and generate datasets, including popular datasets

4) train_test_split is used to split a dataset into two subsets: training and testing datasets. It helps in evaluating the performance of a machine learning model by providing data that the model hasn't seen before (test data).

5) LogisticRegression is a machine learning model used for binary or multi-class classification problems.

6) accuracy_score computes the accuracy of a classification model, which is the ratio of correctly predicted instances to the total instances in the dataset.

In [ ]:
import numpy as np
import pandas as pd
import sklearn.datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [ ]:
# loading the data set from sklearn
#ds=breast cancer dataset
ds=sklearn.datasets.load_breast_cancer()

In [ ]:
#the data is in the form of dictionary as key value pair
print(ds)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [ ]:
# loading the data to a data frame
df=pd.DataFrame(ds.data,columns=ds.feature_names)

In [ ]:
# print the first 5 rows of the dataset
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [ ]:
df.columns

Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', 'worst texture', 'worst perimeter', 'worst area',
       'worst smoothness', 'worst compactness', 'worst concavity',
       'worst concave points', 'worst symmetry', 'worst fractal dimension'],
      dtype='object')

In [ ]:
# adding the 'target' column to the dataset as B/M
df['B/M']=ds.target

In [ ]:
# reading the last 5 rows of the dataframe
df.tail()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  B/M  
564                0.2216          0.2060                  0.07115    0  
565                0.1628          0.2572                  0.06637    0  
566                0.1418          0.2218                  0.07820    0  
567                0.2650          0.4087                  0.12400    0  
568                0.0000          0.2871                  0.07039    1  

[5 rows x 31 columns]

In [ ]:
# reading the first 5 rows of the dataframe
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  B/M  
0          0.4601                  0.11890    0  
1          0.2750                  0.08902    0  
2          0.3613                  0.08758    0  
3          0.6638                  0.17300    0  
4          0.2364                  0.07678    0  

[5 rows x 31 columns]

In [ ]:
# finding the number of rows and columns in the dataset
df.shape

(569, 31)

In [ ]:
#information about data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [ ]:
#checking for the missing values if any
df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
B/M                        0
dtype: int64

In [ ]:
#statistical measures about the data set

df.describe().T

count        mean         std         min  \
mean radius              569.0   14.127292    3.524049    6.981000   
mean texture             569.0   19.289649    4.301036    9.710000   
mean perimeter           569.0   91.969033   24.298981   43.790000   
mean area                569.0  654.889104  351.914129  143.500000   
mean smoothness          569.0    0.096360    0.014064    0.052630   
mean compactness         569.0    0.104341    0.052813    0.019380   
mean concavity           569.0    0.088799    0.079720    0.000000   
mean concave points      569.0    0.048919    0.038803    0.000000   
mean symmetry            569.0    0.181162    0.027414    0.106000   
mean fractal dimension   569.0    0.062798    0.007060    0.049960   
radius error             569.0    0.405172    0.277313    0.111500   
texture error            569.0    1.216853    0.551648    0.360200   
perimeter error          569.0    2.866059    2.021855    0.757000   
area error               569.0   40.337079   45.491006    6.802000   
smoothness error         569.0    0.007041    0.003003    0.001713   
compactness error        569.0    0.025478    0.017908    0.002252   
concavity error          569.0    0.031894    0.030186    0.000000   
concave points error     569.0    0.011796    0.006170    0.000000   
symmetry error           569.0    0.020542    0.008266    0.007882   
fractal dimension error  569.0    0.003795    0.002646    0.000895   
worst radius             569.0   16.269190    4.833242    7.930000   
worst texture            569.0   25.677223    6.146258   12.020000   
worst perimeter          569.0  107.261213   33.602542   50.410000   
worst area               569.0  880.583128  569.356993  185.200000   
worst smoothness         569.0    0.132369    0.022832    0.071170   
worst compactness        569.0    0.254265    0.157336    0.027290   
worst concavity          569.0    0.272188    0.208624    0.000000   
worst concave points     569.0    0.114606    0.065732    0.000000   
worst symmetry           569.0    0.290076    0.061867    0.156500   
worst fractal dimension  569.0    0.083946    0.018061    0.055040   
B/M                      569.0    0.627417    0.483918    0.000000   

                                25%         50%          75%         max  
mean radius               11.700000   13.370000    15.780000    28.11000  
mean texture              16.170000   18.840000    21.800000    39.28000  
mean perimeter            75.170000   86.240000   104.100000   188.50000  
mean area                420.300000  551.100000   782.700000  2501.00000  
mean smoothness            0.086370    0.095870     0.105300     0.16340  
mean compactness           0.064920    0.092630     0.130400     0.34540  
mean concavity             0.029560    0.061540     0.130700     0.42680  
mean concave points        0.020310    0.033500     0.074000     0.20120  
mean symmetry              0.161900    0.179200     0.195700     0.30400  
mean fractal dimension     0.057700    0.061540     0.066120     0.09744  
radius error               0.232400    0.324200     0.478900     2.87300  
texture error              0.833900    1.108000     1.474000     4.88500  
perimeter error            1.606000    2.287000     3.357000    21.98000  
area error                17.850000   24.530000    45.190000   542.20000  
smoothness error           0.005169    0.006380     0.008146     0.03113  
compactness error          0.013080    0.020450     0.032450     0.13540  
concavity error            0.015090    0.025890     0.042050     0.39600  
concave points error       0.007638    0.010930     0.014710     0.05279  
symmetry error             0.015160    0.018730     0.023480     0.07895  
fractal dimension error    0.002248    0.003187     0.004558     0.02984  
worst radius              13.010000   14.970000    18.790000    36.04000  
worst texture             21.080000   25.410000    29.720000    49.54000  
worst perimeter           84.110000   97.660000   125.40000

In [ ]:
#finding the value counts
df['B/M'].value_counts()

B/M
1    357
0    212
Name: count, dtype: int64

Here

1 --> benign

0 --> malignant

In [ ]:
# grouping by data according to B/M column and computing mean values
df.groupby('B/M').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
B/M                                                                           
0      17.462830     21.604906      115.365377  978.376415         0.102898   
1      12.146524     17.914762       78.075406  462.790196         0.092478   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
B/M                                                                         
0            0.145188        0.160775             0.087990       0.192909   
1            0.080085        0.046058             0.025717       0.174186   

     mean fractal dimension  ...  worst radius  worst texture  \
B/M                          ...                                
0                  0.062680  ...     21.134811      29.318208   
1                  0.062867  ...     13.379801      23.515070   

     worst perimeter   worst area  worst smoothness  worst compactness  \
B/M                                                                      
0         141.370330  1422.286321          0.144845           0.374824   
1          87.005938   558.899440          0.124959           0.182673   

     worst concavity  worst concave points  worst symmetry  \
B/M                                                          
0           0.450606              0.182237        0.323468   
1           0.166238              0.074444        0.270246   

     worst fractal dimension  
B/M                           
0                   0.091530  
1                   0.079442  

[2 rows x 30 columns]


Here it is  clearly visible that the average values of malignant tumor cells is higher then the benign tumor cells.

Now we will seperate out dataframe into x and y as feature and target(B/M).

In [ ]:
print(df)

     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mea

In [ ]:
x = df.drop(columns='B/M',axis=1)
y = df['B/M']

In [ ]:
print(x)

     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mea

In [ ]:
print(y)

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: B/M, Length: 569, dtype: int64


In [ ]:
#Now Splitting the data into training and testing data
x_train,x_test,y_train,y_test=train_test_split(x,y,train_size=0.2,random_state=2)

In [ ]:
print(x.shape,x_train.shape,x_test.shape)

(569, 30) (113, 30) (456, 30)


#Applying Logistic Regression

In [ ]:
model = LogisticRegression()

In [ ]:
# Training the logistic regression linear model using the training data

model.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

The above classification(logistic regression)will try to find out the relationship between all the x values and the y values. This means that it will analyse how the data is when the label is 0 and how the data is when the label is 1.

Now we will evaluate the model's Accuracy

In [ ]:
# accuracy on training data

x_train_prediction = model.predict(x_train)
training_data_accuracy = accuracy_score(y_train,x_train_prediction)

print('Accuracy on training data :',training_data_accuracy)

Accuracy on training data : 0.9823008849557522


This shows that 98% of the label values predicted from the training data are correct.


In [ ]:
# accuracy on testing data

x_test_prediction = model.predict(x_test)
testing_data_accuracy = accuracy_score(y_test,x_test_prediction)

print('Accuracy on testing data :',testing_data_accuracy)

Accuracy on testing data : 0.9254385964912281


This shows that 93% of the label values predicted from the test data are correct. Usually test data accuracy is much less than training data accuracy.

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import sklearn.datasets

breast cancer prediction with pytorch

In [ ]:
#check for CUDA availability
device=torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device:{device}")

Using device:cpu


In [ ]:

X=ds.data
#df.data contains the features of the dataset.
y=ds.target

In [ ]:
print(X)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]


In [ ]:

print(y[:5])

[0 0 0 0 0]


In [ ]:
X_train ,X_test, y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=+42)

In [ ]:
print(X.shape)
print(X_train.shape)
print(X_test.shape)

(569, 30)
(455, 30)
(114, 30)


In [ ]:
scaler= StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

In [ ]:
type(X_train)

numpy.ndarray

In [ ]:
X_train = torch.tensor(X_train, dtype=torch.float32).to(device)

In [ ]:
y_train = torch.tensor(y_train, dtype=torch.float32).to(device)

In [ ]:
X_test = torch.tensor(X_test, dtype=torch.float32).to(device)

In [ ]:
y_test = torch.tensor(y_test, dtype=torch.float32).to(device)

In [ ]:
y_test

tensor([1., 0., 0., 1., 1., 0., 0., 0., 1., 1., 1., 0., 1., 0., 1., 0., 1., 1.,
        1., 0., 0., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1.,
        0., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 1., 1.,
        1., 1., 1., 0., 0., 1., 1., 0., 0., 1., 1., 1., 0., 0., 1., 1., 0., 0.,
        1., 0., 1., 1., 1., 0., 1., 1., 0., 1., 0., 0., 0., 0., 0., 0., 1., 1.,
        1., 1., 1., 1., 1., 1., 0., 0., 1., 0., 0., 1., 0., 0., 1., 1., 1., 0.,
        1., 1., 0., 1., 1., 0.])

In [ ]:
y_test = y_test.clone().detach().float().to(device)


In [ ]:
y_test

tensor([1., 0., 0., 1., 1., 0., 0., 0., 1., 1., 1., 0., 1., 0., 1., 0., 1., 1.,
        1., 0., 0., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1.,
        0., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 1., 1.,
        1., 1., 1., 0., 0., 1., 1., 0., 0., 1., 1., 1., 0., 0., 1., 1., 0., 0.,
        1., 0., 1., 1., 1., 0., 1., 1., 0., 1., 0., 0., 0., 0., 0., 0., 1., 1.,
        1., 1., 1., 1., 1., 1., 0., 0., 1., 0., 0., 1., 0., 0., 1., 1., 1., 0.,
        1., 1., 0., 1., 1., 0.])

In [ ]:

class NeuralNet(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(NeuralNet, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, output_size)
        self.sigmoid = nn.Sigmoid()

    def forward(self, X):  # pass the data to this forward function
        out = self.fc1(X)
        out = self.relu(out)
        out = self.fc2(out)
        out = self.sigmoid(out)
        return out

In [ ]:
# Define hyperparameters
input_size = X_train.shape[1]
hidden_size = 64
output_size = 1
learning_rate = 0.001
num_epochs = 100

In [ ]:
#initialize the neural network and move it to GPU
model=NeuralNet(input_size,hidden_size,output_size).to(device)

In [ ]:
#define loss and the optilizer
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

In [ ]:

for epoch in range(num_epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)  # No need to call forward explicitly
    loss = criterion(outputs, y_train.view(-1, 1))
    loss.backward()
    optimizer.step()

    # Calculate accuracy
    with torch.no_grad():
        predicted = outputs.round()
        correct = (predicted == y_train.view(-1, 1)).float().sum()
        accuracy = correct / y_train.size(0)

    if (epoch + 1) % 10 == 0:
        print(f"Epoch {epoch + 1}/{num_epochs}, Loss: {loss.item():.4f}, Accuracy: {accuracy.item() * 100:.2f}%")

Epoch 10/100, Loss: 0.5692, Accuracy: 90.11%
Epoch 20/100, Loss: 0.4351, Accuracy: 92.75%
Epoch 30/100, Loss: 0.3374, Accuracy: 93.41%
Epoch 40/100, Loss: 0.2674, Accuracy: 93.85%
Epoch 50/100, Loss: 0.2184, Accuracy: 95.60%
Epoch 60/100, Loss: 0.1843, Accuracy: 96.26%
Epoch 70/100, Loss: 0.1600, Accuracy: 96.92%
Epoch 80/100, Loss: 0.1421, Accuracy: 97.14%
Epoch 90/100, Loss: 0.1284, Accuracy: 97.14%
Epoch 100/100, Loss: 0.1174, Accuracy: 97.36%


In [ ]:
#evaluation on training set
model.eval()
with torch.no_grad():
  outputs=model(X_train)
  predicted=outputs.round()
  correct=(predicted==y_train.view(-1,1)).float().sum()  # Change here
  accuracy=correct/y_train.size(0)
  print(f"accuracy on training data:{accuracy.item()*100:.2f}%")
  # Reshape y_train to match the predicted output

accuracy on training data:97.36%


In [ ]:
#evaluation on training set
model.eval()
with torch.no_grad():
  outputs=model(X_test)
  predicted=outputs.round()
  correct=(predicted==y_test.view(-1,1)).float().sum()  # Change here
  accuracy=correct/y_test.size(0)
  print(f"accuracy on testing data:{accuracy.item()*100:.2f}%")

accuracy on testing data:98.25%


LR model achieved a high
training accuracy of 98%, demonstrating its ability to fit the
training data well. It performed well on the testing data,
achieving an accuracy of 93%. This indicates that the model
generalizes reasonably well to unseen data.

MLP achieved a training accuracy
of 97.3%. This suggests that the network was effective
in learning from the training data, capturing the important
patterns required for classification. The model has slightly
higher testing accuracy of 98.25%, which indicates excellent
generalization to new data. The high accuracy on both training
and testing datasets demonstrates the model’s robustness and
reliability in predicting breast cancer classifications.

MLP demonstrated superior performance over the Logistic
Regression model, especially in terms of testing accuracy. This
suggests that the MLP is more capable of capturing complex
patterns in the data which makes it a more effective tool for
breast cancer classification.